### PR TITLE
TableRoller redesign

### PR DESCRIPTION
**Pull request by JYCasalis**
- Redesign TableRoller not to use recursive rendering of embedded
  sub rollers but, instead, execute them and consolidate the result
  (and the execution steps in the tooltip).
- Support multiple rolls in lookup case too (was only available for
  non lookup case).
- Address async issue with GenericFileRoller

See issue #116

### CODE REVIEW

File: src/assets/main.css [hunk 1]
```diff
@@ -13,9 +13,9 @@
 .dice-roller > div {
     display: inline-block;
 }
-.dice-roller span {
+/* .dice-roller span {
     font-weight: 700;
-}
+} */
```
Comment (valentine195) [inline src/assets/main.css:18]: I'm not sure I like the idea of just blanket removing the bold. Maybe make this an option in settings, or target embedded rollers specifically?

File: src/main.ts [hunk 2]
```diff
@@ -703,13 +703,15 @@ export default class DiceRollerPlugin extends Plugin {
...
             case "table": {
-                return new TableRoller(
+                const roller = new TableRoller(
                     this,
                     content,
                     lexemes[0],
                     source,
                     showDice
                 );
+                await roller.initPromise;
```
Comment (valentine195) [inline src/main.ts:713]: Is there a reason for making `this.initPromise` vs. just directly calling `this.getFile()`?

File: src/roller/table.ts [hunk 3]
```diff
@@ -38,40 +43,148 @@ export class TableRoller extends GenericFileRoller<string> {
...
+
+    async getSubResult(input: string): Promise<SubRollerResult> {
+        let res: SubRollerResult = new SubRollerResult();
+        res.result = input;
+
+        let subTooltips: string[] = [];
+
+        // WARN: we may receive an input that is not string (but a number). Check
+        // for embeded formulas only if we can.
+        if (typeof input?.matchAll === "function") {
```
Comment (valentine195) [inline src/roller/table.ts:100]: You can test directly if this is a string:

`typeof input == "string"`

File: src/roller/table.ts [hunk 4]
```diff
@@ -38,40 +43,148 @@ export class TableRoller extends GenericFileRoller<string> {
...
+                output += tab.repeat(tabCount);
+            }
+            else {
+                output += input.charAt(i);
+            }
+        }
+        return output;
+    }
+
+    async getSubResult(input: string): Promise<SubRollerResult> {
```
Comment (valentine195) [inline src/roller/table.ts:92]: Below you say that input might be a number. I would suggest making that explicit in the typing.

File: src/roller/table.ts [hunk 5]
```diff
@@ -38,40 +43,148 @@ export class TableRoller extends GenericFileRoller<string> {
...
+                );
+                if (option) {
+                    subTooltip = this.lookupRoller.original.trim() + " --> " + `${this.lookupRoller.resultText}${this.header ? " | " + this.header : ""}`.trim();
+                    selectedOption = option[1];
+                }
+            }
+            else {
+                const options = [...this.options];
+                const randomRowNumber = this.getRandomBetween(0, options.length - 1);
+                // TODO: to be confirmed (was this to forbid rolling the same result twice ?)
```
Comment (valentine195) [inline src/roller/table.ts:159]: I believe that was the case, but I actually don't remember lol

File: src/assets/main.css (hunk 1 above)
Comment (JYCasalis) [reply src/assets/main.css:18]: You're right!
I did that change based on one example provided in another issue where the output was:

`**Gender**: `dice; [[Note^Table]]``

With the bold coming from the .dice-roller rule, this wasn't looking good (to my eyes at least).

But... it's better to keep the bold by default and let users who want to do something like above to add their own css snippet to remove it.

So... all of this to say: I'll remove this change.

File: src/roller/table.ts (hunk 3 above)
Comment (JYCasalis) [reply src/roller/table.ts:100]: Yeah, I agree that was too much!
Will change it.

File: src/roller/table.ts (hunk 4 above)
Comment (JYCasalis) [reply src/roller/table.ts:92]: Another one I wanted to discuss (I'm not familiar enough with typescript).

I have been surprised to realize that: 

```
function f(input: string) {
    console.log(typeof input);
}
const n:number = 12;
f(n);
```
is considered valid and ... displays "number" in the console!

In fact I **don't** want to allow calling getSubResult with a number, just with a string.

If your input is a number, then convert it to a string before calling me.
I thought that **input:string** would "guarantee" that but... that's not the case (I can call it with any type apparently).
(yeah, I know ... typescript is not C++ :-p)

Not really sure what to do in fact!

File: src/roller/table.ts (hunk 5 above)
Comment (JYCasalis) [reply src/roller/table.ts:159]: OK so I'll remove it; no reason not to be able to get twice the same result for me ;-)

File: src/roller/table.ts [hunk 6]
```diff
@@ -38,40 +43,148 @@ export class TableRoller extends GenericFileRoller<string> {
...
-        return [...Array(this.rolls)]
-            .map(() => {
-                let option =
-                    options[this.getRandomBetween(0, options.length - 1)];
-                options.splice(options.indexOf(option), 1);
-                return option;
-            })
-            .join("||");
+
+        // TODO: find a simpler way
```
[the remark this answers is not in the document]
Comment (JYCasalis) [reply src/roller/table.ts:174]: Hum, not sure that nested arrays are simpler!

I suppose that you would have an array for each "level", but then... don't forget that each level would need potentially a different "join character".

I'm not sure I see what you have in mind.

[the remark this answers is not in the document]
Comment (JYCasalis) [reply src/roller/table.ts:174]: As for the looping per character, this is where I think it could be "optimized" using some regex to "consume" all characters that are of no interest for the formatting in one shot).

[the remark this answers is not in the document]
Comment (JYCasalis) [reply src/roller/table.ts:174]: OK, so I tried using regex... not sure it looks way better though 😁

```
prettify(input: string): string {
        const specialChars = /(.*?)(\(|\)|;|\|\|)(.*)/
        const tab = "\t";
        let tabCount = 0;
        let output:string = "";

        let remaining:string = input;
        let matches: RegExpMatchArray;
        while (matches = remaining.match(specialChars)) {
            let [, beforeSpecial, special, afterSpecial] = matches;
            output += beforeSpecial;
            if (special == ")") {
                tabCount--;
                output += "\n";
                output += tab.repeat(tabCount);
                output += ")";
            }
            else {
                if (special == "(") {
                    tabCount++;
                    output += "(";
                }
                else if (special == ";") {
                    output += ",";
                }
                else if (special == "||") {
                    output += "|";
                }
                output += "\n"
                output += tab.repeat(tabCount);
            }
            remaining = afterSpecial;
        }
        output += remaining;

        return output;
    }
```

File: src/roller/table.ts (hunk 4 above)
Comment (valentine195) [reply src/roller/table.ts:95]: Yep... typescript is just syntactic sugar. The types have no bearing on what is actually generated in the code (outside of `enums` but that's a separate thing).

All you're doing is telling the typescript language service that the input is a string. In practice you can call the method with anything... the idea is that by telling the language service "this is only ever a string", you will receive an error message from the language service if you call the method with a parameter that is _not_ a string, and you can take it a step further and make the build fail on these types of errors. _But once the JavaScript is generated, it doesn't matter anymore._

Basically my suggestion is to say `input: string | number` and then the language service will show you if there is a possibility that passing in a number will throw an error.

Comment (JYCasalis) [reply src/roller/table.ts:95]: OK, I know what I wasn't understanding.

getSubResult is called with subRoller.result which is a BasicRoller where result is... `any` and as soon as there's an any, all type checking are "ignored".

Anyway, I'll do as you suggested (even if that hurts my brain :-p)

**Comment by JYCasalis** [comment]
Just added another fix / optimization on top linked to what HAJ523 reported.
Now dice-roller reads and updates a file only once whatever the number of dice-mod blocks in the note (previously the number of read/write was... factorial(nb dice-mod blocks) 😂 ) .

**Comment by valentine195** [comment]
LGTM
